src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,907 @@
+import { EventEmitter } from 'events';
+import HDKey from 'hdkey';
+import * as ethUtil from 'ethereumjs-util';
+import * as sigUtil from 'eth-sig-util';
+import { TransactionFactory, TxData, TypedTransaction } from '@ethereumjs/tx';
+import type OldEthJsTransaction from 'ethereumjs-tx';
+import type LedgerHwAppEth from '@ledgerhq/hw-app-eth';
+
+const pathBase = 'm';
+const hdPathString = `${pathBase}/44'/60'/0'`;
+const keyringType = 'Ledger Hardware';
+
+const BRIDGE_URL = 'https://metamask.github.io/eth-ledger-bridge-keyring';
+
+const MAX_INDEX = 1000;
+const NETWORK_API_URLS: Record<string, string> = {
+  ropsten: 'http://api-ropsten.etherscan.io',
+  kovan: 'http://api-kovan.etherscan.io',
+  rinkeby: 'https://api-rinkeby.etherscan.io',
+  mainnet: 'https://api.etherscan.io',
+} as const;
+
+enum IFrameMessageAction {
+  LedgerConnectionChange = 'ledger-connection-change',
+  LedgerUnlock = 'ledger-unlock',
+  LedgerMakeApp = 'ledger-make-app',
+  LedgerUpdateTransport = 'ledger-update-transport',
+  LedgerSignTransaction = 'ledger-sign-transaction',
+  LedgerSignPersonalMessage = 'ledger-sign-personal-message',
+  LedgerSignTypedData = 'ledger-sign-typed-data',
+}
+
+interface IFrameMessage {
+  action: IFrameMessageAction;
+  params: Record<string, unknown>;
+  target: string;
+  messageId: number;
+}
+
+type GetAddressPayload = Awaited<ReturnType<LedgerHwAppEth['getAddress']>>;
+
+type SignMessagePayload = Awaited<
+  ReturnType<LedgerHwAppEth['signEIP712HashedMessage']>
+>;
+
+type SignTransactionPayload = Awaited<
+  ReturnType<LedgerHwAppEth['signTransaction']>
+>;
+
+interface ConnectionChangedPayload {
+  connected: boolean;
+}
+
+type IFrameMessageResponsePayload = { error?: unknown } & (
+  | GetAddressPayload
+  | SignTransactionPayload
+  | SignMessagePayload
+  | ConnectionChangedPayload
+);
+
+interface IFrameMessageResponse {
+  success: boolean;
+  action: IFrameMessageAction;
+  messageId: number;
+  payload: IFrameMessageResponsePayload;
+  error?: unknown;
+}
+
+export interface AccountDetails {
+  index?: number;
+  bip44?: boolean;
+  hdPath?: string;
+}
+
+export interface LedgerBridgeKeyringOptions {
```

Comment:
Nit: if all these fields are marked as optional with `?:` then maybe it would be better to remove the optionality (i.e., remove the `?:` and replace it with `:`, and just replace references to `LedgerBridgeKeyringOptions` with `Partial<LedgerBridgeKeyringOptions>`?

IMO this makes more sense as a person _likely_ shouldn't be instantiating the bridge with something like:

```ts
foo({
  hdPath: undefined,
  accounts: ['foo', 'bar'],
  implementFullBIP44: undefined
})